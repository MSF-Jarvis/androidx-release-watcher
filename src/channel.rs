use semver::Version;
use std::cmp::{Eq, PartialEq, PartialOrd};
use std::convert::TryFrom;
use std::fmt::Debug;
use std::str::FromStr;
use thiserror::Error;

/// Release channels for AndroidX packages
/// Since we're deriving [PartialOrd] automatically, the order
/// of these fields is crucial. Sort by stability, not alphabetical
/// order.
#[derive(Debug, Eq, PartialEq, PartialOrd)]
pub(crate) enum Channel {
    Dev,
    Alpha,
    Beta,
    RC,
    Stable,
}

#[derive(Debug, Error)]
pub(crate) enum ChannelError {
    #[error("no match found")]
    NoMatchFound,
    #[error("failed to determine channel for {0}")]
    FailedToParseVersion(Version),
}

impl FromStr for Channel {
    type Err = ChannelError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "alpha" => Ok(Channel::Alpha),
            "beta" => Ok(Channel::Beta),
            "dev" => Ok(Channel::Dev),
            "rc" => Ok(Channel::RC),
            "stable" => Ok(Channel::Stable),
            _ => Err(ChannelError::NoMatchFound),
        }
    }
}

impl TryFrom<Version> for Channel {
    type Error = ChannelError;

    fn try_from(value: Version) -> Result<Self, Self::Error> {
        if value.pre.is_empty() {
            return Ok(Channel::Stable);
        };
        if let Some(pre) = value.pre.get(0) {
            let pre_str = pre.to_string();
            return if pre_str.starts_with("alpha") {
                Ok(Channel::Alpha)
            } else if pre_str.starts_with("beta") {
                Ok(Channel::Beta)
            } else if pre_str.starts_with("dev") {
                Ok(Channel::Dev)
            } else if pre_str.starts_with("rc") {
                Ok(Channel::RC)
            } else {
                Err(ChannelError::FailedToParseVersion(value))
            };
        } else {
            return Ok(Channel::Stable);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Channel;
    use semver::{AlphaNumeric, Version};
    use std::convert::TryFrom;

    #[test]
    fn alpha_version() {
        let v = Version {
            major: 1,
            minor: 1,
            patch: 0,
            pre: vec![AlphaNumeric("alpha01".to_owned())],
            build: vec![],
        };
        let channel = Channel::try_from(v);
        assert_eq!(Channel::Alpha, channel.unwrap());
    }

    #[test]
    fn beta_version() {
        let v = Version {
            major: 1,
            minor: 1,
            patch: 0,
            pre: vec![AlphaNumeric("beta01".to_owned())],
            build: vec![],
        };
        let channel = Channel::try_from(v);
        assert_eq!(Channel::Beta, channel.unwrap());
    }

    #[test]
    fn dev_version() {
        let v = Version {
            major: 1,
            minor: 1,
            patch: 0,
            pre: vec![AlphaNumeric("dev01".to_owned())],
            build: vec![],
        };
        let channel = Channel::try_from(v);
        assert_eq!(Channel::Dev, channel.unwrap());
    }

    #[test]
    fn rc_version() {
        let v = Version {
            major: 1,
            minor: 1,
            patch: 0,
            pre: vec![AlphaNumeric("rc01".to_owned())],
            build: vec![],
        };
        let channel = Channel::try_from(v);
        assert_eq!(Channel::RC, channel.unwrap());
    }

    #[test]
    fn stable_version() {
        let v = Version::new(1, 1, 1);
        let channel = Channel::try_from(v);
        assert_eq!(Channel::Stable, channel.unwrap());
    }

    #[test]
    fn cmp_channels() {
        assert!(Channel::Stable > Channel::RC);
        assert!(Channel::RC > Channel::Beta);
        assert!(Channel::Beta > Channel::Alpha);
        assert!(Channel::Alpha > Channel::Dev);
    }
}
