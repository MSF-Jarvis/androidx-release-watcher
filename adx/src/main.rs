mod channel;
mod package;
mod parse;
#[cfg(feature = "measure-alloc")]
mod stats_alloc;

use channel::Channel;
use clap::{crate_authors, crate_description, crate_name, crate_version, AppSettings, Clap};
use color_eyre::Result;

#[cfg(feature = "measure-alloc")]
use stats_alloc::{Region, StatsAlloc, INSTRUMENTED_SYSTEM};
#[cfg(feature = "measure-alloc")]
use std::alloc::System;

#[cfg(feature = "measure-alloc")]
#[global_allocator]
static GLOBAL: &StatsAlloc<System> = &INSTRUMENTED_SYSTEM;

#[derive(Clap)]
#[clap(
    name = crate_name!(),
    version = crate_version!(),
    author = crate_authors!(),
    about = crate_description!(),
    setting = AppSettings::ColoredHelp,
    setting = AppSettings::DeriveDisplayOrder,
)]
pub(crate) struct Cli {
    /// search term to filter packages with
    #[clap(required = true)]
    pub(crate) search_term: String,
    /// the release channel to find packages from
    #[clap(short='c', long="channel", possible_values=&["dev", "alpha", "beta", "rc", "stable"], default_value="alpha")]
    pub(crate) channel: Channel,
}

#[async_std::main]
async fn main() -> Result<()> {
    color_eyre::install()?;
    #[cfg(feature = "measure-alloc")]
    let reg = Region::new(GLOBAL);
    let cli = Cli::parse();
    let packages = crate::parse::parse(&cli.search_term, cli.channel).await?;
    if packages.is_empty() {
        println!("No results found!");
    } else {
        for package in packages.iter() {
            println!("{}", package);
        }
    };
    #[cfg(feature = "measure-alloc")]
    println!("Stats at end: {:#?}", reg.change());
    Ok(())
}
