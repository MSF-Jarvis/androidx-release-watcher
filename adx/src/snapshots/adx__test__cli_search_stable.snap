---
source: adx/src/main.rs
info:
  program: adx
  args:
    - "--channel"
    - stable
    - appcompat
---
success: true
exit_code: 0
----- stdout -----
androidx.appcompat:appcompat:1.6.1
androidx.appcompat:appcompat-resources:1.6.1

----- stderr -----
